pyth-sdk/src/price.rs

```diff
@@ -897,4 +1201,966 @@ mod test {
+            i64::MAX-100,
+            -9
+        );
+        // fails bc 0-i64::MIN > i64::MAX, so overflow
+        fails(
+            i64::MIN,
+            pc(100, 0, -9),
+            0,
+            pc(0, 0, -9),
+            0,
+            -9
+        );
+    }
+
+    // quickcheck to confirm affine_combination introduces no error if normalization done explicitly on prices first
+    #[quickcheck]
+    fn quickcheck_affine_combination_normalize_prices(x1_inp: i32, y1: Price, x2_inp: i32, y2: Price, x_query_inp: i32) -> TestResult {
+        let x1 = i64::try_from(x1_inp).ok().unwrap();
+        let x2 = i64::try_from(x2_inp).ok().unwrap();
+        let x_query = i64::try_from(x_query_inp).ok().unwrap();
+
+        let pre_add_expo = -9;
+
+        // require x2 > x1
+        if x1 >= x2 {
+            return TestResult::discard()
+        }
+
+        // require low pre_add_expo
+        if pre_add_expo >= 2 {
+            return TestResult::discard()
+        }
+
+        // require reasonable price/conf range
+        if (y1.price > 2*MAX_PD_V_I64) || (y1.price < 2*MIN_PD_V_I64) {
+            return TestResult::discard()
+        }
+        if (y2.price > 2*MAX_PD_V_I64) || (y2.price < 2*MIN_PD_V_I64) {
+            return TestResult::discard()
+        }
+
+        let result_orig = Price::affine_combination(x1, y1, x2, y2, x_query, pre_add_expo).unwrap();
+
+        let y1_norm = y1.normalize().unwrap();
+        let y2_norm = y2.normalize().unwrap();
+
+        let result_norm = Price::affine_combination(x1, y1_norm, x2, y2_norm, x_query, pre_add_expo).unwrap();
+
+        TestResult::from_bool(result_norm == result_orig)
+    }
+
+    // quickcheck to confirm affine_combination introduces bounded error if close fraction x/y passed in first
+    #[quickcheck]
+    fn quickcheck_affine_combination_normalize_fractions(x1_inp: i32, y1: Price, x2_inp: i32, y2: Price, x_query_inp: i32) -> TestResult {
+        let x1 = i64::try_from(x1_inp).ok().unwrap();
+        let x2 = i64::try_from(x2_inp).ok().unwrap();
+        let x_query = i64::try_from(x_query_inp).ok().unwrap();
+
+        let pre_add_expo = -9;
+
+        // require x2 > x1
+        if x1 >= x2 {
+            return TestResult::discard()
+        }
+
+        if (x_query > 5*x2) || (x_query < 2*x1 - x2) {
+            return TestResult::discard()
+        }
+
+        // require low pre_add_expo
+        if pre_add_expo >= 2 {
+            return TestResult::discard()
+        }
+
+        // require reasonable price/conf range
+        if (y1.price > 2*MAX_PD_V_I64) || (y1.price < 2*MIN_PD_V_I64) {
+            return TestResult::discard()
+        }
+        if (y2.price > 2*MAX_PD_V_I64) || (y2.price < 2*MIN_PD_V_I64) {
+            return TestResult::discard()
+        }
+
+        let x1_new: i64;
+        let xq_new: i64;
+        let x2_new: i64;
+
+        if x2 == 0 {
+            x1_new = x1;
+            xq_new = x_query;
+            x2_new = x2;
+        }
```

Comment:
not sure i understand this if statement. isn't the bad case for fraction where x2 - x1 == 0?

Reply:
that's checked for earlier